Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m

```diff
@@ -359,72 +353,124 @@ - (void)configureViews
 
         bodyMediaView.layoutMargins = UIEdgeInsetsZero;
         [self.viewConstraints addObjectsFromArray:@[
-            [footerView autoPinLeadingToSuperviewMarginWithInset:self.conversationStyle.textInsetHorizontal],
-            [footerView autoPinTrailingToSuperviewMarginWithInset:self.conversationStyle.textInsetHorizontal],
-            [footerView autoPinBottomToSuperviewMarginWithInset:self.conversationStyle.textInsetBottom],
+            [self.footerView autoPinLeadingToSuperviewMarginWithInset:self.conversationStyle.textInsetHorizontal],
+            [self.footerView autoPinTrailingToSuperviewMarginWithInset:self.conversationStyle.textInsetHorizontal],
+            [self.footerView autoPinBottomToSuperviewMarginWithInset:self.conversationStyle.textInsetBottom],
         ]];
-        [self.footerView setHasShadows:YES viewItem:self.viewItem];
     } else {
-        OWSFail(@"%@ could not display footer.", self.logTag);
+        // Display footer at bottom of message bubble.
+        [self.footerView configureWithConversationViewItem:self.viewItem hasShadows:NO];
+        [textViews addObject:self.footerView];
     }
 
-    if (textStackView) {
-        CGSize bubbleSize = [self measureSize];
-        [NSLayoutConstraint autoSetPriority:UILayoutPriorityRequired
-                             forConstraints:^{
-                                 [self.viewConstraints addObjectsFromArray:@[
-                                     [self autoSetDimension:ALDimensionWidth toSize:bubbleSize.width],
-                                 ]];
-                             }];
+    [self insertTextViewsIntoStackViewIfNecessary:textViews];
+
+    CGSize bubbleSize = [self measureSize];
+    [NSLayoutConstraint autoSetPriority:UILayoutPriorityRequired
+                         forConstraints:^{
+                             [self.viewConstraints addObjectsFromArray:@[
+                                 [self autoSetDimension:ALDimensionWidth toSize:bubbleSize.width],
+                             ]];
+                         }];
+
+    [self updateBubbleColorWithBodyMediaView:bodyMediaView];
+}
+
+- (void)updateBubbleColorWithBodyMediaView:(nullable UIView *)bodyMediaView
+{
+    BOOL hasOnlyBodyMediaView = NO;
+    switch (self.cellType) {
+        case OWSMessageCellType_Unknown:
+        case OWSMessageCellType_TextMessage:
+        case OWSMessageCellType_OversizeTextMessage:
+        case OWSMessageCellType_GenericAttachment:
+        case OWSMessageCellType_DownloadingAttachment:
+        case OWSMessageCellType_ContactShare:
+            break;
+        case OWSMessageCellType_StillImage:
+        case OWSMessageCellType_AnimatedImage:
+        case OWSMessageCellType_Audio:
+        case OWSMessageCellType_Video:
+            hasOnlyBodyMediaView = (bodyMediaView && self.stackView.subviews.count == 1);
+            break;
+    }
+    if ([self.viewItem.interaction isKindOfClass:[TSMessage class]] && !hasOnlyBodyMediaView) {
```

Comment:
In what case would self.viewItem.interaction not be a TSMessage? If none, a guard pattern/owsFail would be clearer.

Reply:
Removed.